pkg/dns/client/dns.go

```diff
@@ -396,15 +402,59 @@ func (h *LocalDNSServer) queryUpstream(upstreamClient *dns.Client, req *dns.Msg,
 		}
 		scope.Infof("upstream failure: %v", err)
 	}
+
 	if response == nil {
-		failures.Increment()
-		response = new(dns.Msg)
-		response.SetReply(req)
-		response.Rcode = dns.RcodeServerFailure
+		response = serverFailure(req)
 	}
 	return response
 }
 
+// queryUpstreamParallel will send parallel queries to all nameservers and return first response immediately.
+func (h *LocalDNSServer) queryUpstreamParallel(upstreamClient *dns.Client, req *dns.Msg, scope *istiolog.Scope) *dns.Msg {
+	responseCh := make(chan *dns.Msg, 1)
+	errCh := make(chan error, len(h.resolvConfServers))
+
+	queryOne := func(upstream string) {
+		cResponse, _, err := upstreamClient.Exchange(req, upstream)
```

Comment:
should we use ExchangeContext so we can cancel other requests once one completes?

Reply:
I tested it but failed. `ExchangeContext` can't be canceled by context. It just extract deadline from the context.